File: rest-service/src/main/java/com/mdbank/service/PositionService.java
package com.mdbank.service;

import com.mdbank.config.GlobalDataConfig;
import com.mdbank.model.Position;
import com.mdbank.repository.PositionRepository;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.util.StopWatch;

import javax.annotation.PostConstruct;
import java.util.ArrayList;
import java.util.List;

@Component
@Slf4j
public class PositionService {
    final private PositionRepository repository;
    private double minLatitude;
    private double maxLatitude;
    private double minLongitude;
    private double maxLongitude;
    private double latStep;
    private double lonStep;
    @Getter
    private List<Position> positionList;

    @Autowired
    public PositionService(GlobalDataConfig globalDataConfig,
                           PositionRepository repository) {
        this.minLatitude = globalDataConfig.getMinLatitude();
        this.maxLatitude = globalDataConfig.getMaxLatitude();
        this.minLongitude = globalDataConfig.getMinLongitude();
        this.maxLongitude = globalDataConfig.getMaxLongitude();
        this.latStep = globalDataConfig.getLatStep();
        this.lonStep = globalDataConfig.getLonStep();
        this.repository = repository;
    }

    @PostConstruct
    private void init() {
        log.info("Initialization position service started");
        StopWatch timer = new StopWatch();
        timer.start();

        positionList = repository.findAll();
        //если в базе нет данных о координатах, то берём их их конфигурации и потом сохраняем в базе
        if (positionList.isEmpty()) {
            positionList = new ArrayList<>();
            for (double latitude = minLatitude; latitude <= maxLatitude; latitude += latStep) {
                for (double longitude = minLongitude; longitude <= maxLongitude; longitude += lonStep) {
                    Position position = new Position(latitude, longitude);
                    positionList.add(position);
                }
            }
            repository.saveAll(positionList);
        }

        timer.stop();
        log.info("Initialization position service ended. Execution time: {} seconds", timer.getTotalTimeSeconds());
    }

    public Position getPositionWithId(Position position) {
        if (position.getId() == null) {
            return repository.findOptionalByLatitudeAndLongitude(position.getLatitude(), position.getLongitude())
                    .orElseGet(() -> repository.save(position));
        } else {
            return position;
        }
    }
}

File: rest-service/src/main/java/com/mdbank/service/PositionService.kt
package com.mdbank.service

import com.mdbank.config.GlobalDataConfig
import com.mdbank.exception.InitializationException
import com.mdbank.model.Position
import com.mdbank.model.transformIndexToLat
import com.mdbank.model.transformIndexToLon
import com.mdbank.repository.PositionRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.annotation.PostConstruct

@Service
class PositionService @Autowired constructor(globalDataConfig: GlobalDataConfig, private val repository: PositionRepository) {
    private var cachedPositions: MutableList<Position> = ArrayList()
    private var minLatitude: Double = globalDataConfig.minLatitude ?: throw InitializationException()
    private var maxLatitude: Double = globalDataConfig.maxLatitude ?: throw InitializationException()
    private var minLongitude: Double = globalDataConfig.minLongitude ?: throw InitializationException()
    private var maxLongitude: Double = globalDataConfig.maxLongitude ?: throw InitializationException()
    private val latStep: Double = globalDataConfig.latStep
    private val lonStep: Double = globalDataConfig.lonStep
    private val log = LoggerFactory.getLogger(this.javaClass)

    @Transactional
    internal fun initAllPositions() {
        val minPos = Position(latitude = minLatitude, longitude = minLongitude)
        val maxPos = Position(latitude = maxLatitude, longitude = maxLongitude)

        val maxLongIndex = maxPos.longIndex
        val minLongIndex = minPos.longIndex
        val maxLatIndex = maxPos.latIndex
        val minLatIndex = minPos.latIndex

        val count = (maxLongIndex - minLongIndex + 1) * (maxLatIndex - minLatIndex + 1)

        if (count.toLong() != repository.count()) {
            val positions = HashSet<Position>()
            for (longIndex in minLongIndex..maxLongIndex) {
                for (latIndex in minLatIndex..maxLatIndex) {
                    val position = Position(latitude = transformIndexToLat(latIndex), longitude = transformIndexToLon(longIndex))
                    positions.add(position)
                }
            }

            repository.saveAll(positions)
        }
    }

    fun forEachPosition(consumer: (Position) -> Unit) {
        cachedPositions.forEach(consumer)
    }

    @Transactional(readOnly = true)
    fun getPosition(latitude: Double, longitude: Double): Position? {
        return repository.findByLatitudeAndLongitude(latitude, longitude)
    }

    @Transactional(readOnly = true)
    fun getAllPositions(): List<Position> {
        return repository.findAll()
    }
}